wandb/sdk/launch/_project_spec.py

```diff
@@ -74,6 +81,22 @@ def __init__(
 
         self.clear_parameter_run_config_collisions()
 
+    @property
+    def base_image(self) -> str:
+        """Returns {PROJECT}_base:{PYTHON_VERSION}"""
+        # TODO: this should likely be source_project when we have it...
+        generated_name = "{}_base:{}".format(
+            self.target_project, self.python_version or "3"
+        )
+        return self._base_image or generated_name
+
+    @property
+    def image_name(self) -> str:
+        """Returns {PROJECT}_launch the ultimate version will
+        be tagged with a sha of the git repo"""
+        # TODO: this should likely be source_project when we have it...
+        return "{}_launch".format(self.target_project)
```

Comment:
I'm naming the base_image and the ultimate image after the `target_project`. it would probably make more sense to name these after the source project but may not be a big deal.

Reply:
I think we need to include some unique identifier. In the case where a user has several agents running, if they all use the same repo and target project they could end up with collisions between them. Shawn was experiencing this when running many agents simultaneously.

Reply:
Actually, since we move the run id to the container command, I think this is fine.

Reply:
Yeah the tag for the final image will be namespaced by the git commit and the other information should all be coming in from outside.

Reply:
the git commit doesn't seem like it's sufficient disambiguation, you could have multiple agents running the same code or code that differs but isn't committed? why not use the run id

Reply:
We want to avoid rebuilding images.  The base_image will still have code copied into it so it's ok if multiple runs had the same commit.  The final image should likely be tagged with the run id if we aren't already.

Reply:
Actually I need to re-trace the path to be sure this is ok...

Reply:
@stephchen you're totally right, we need to tag the final built image with at-least the source run_id.  It would probably be wise to just use the target run_id if we have it.  That would handle changes to the launch config that effected the build.

Reply:
It would be nice if we could prevent rebuilds for many instances of the same source run.  Maybe we default to source run and let the user specify to rebuild in the launch config...

Reply:
In the case of local runs, we'll need a different unique identifier. For now, I'll use the target run_id in those cases. I'll use the source run id, and add a rebuild flag in the launch config's docker config key.